### ISSUE TITLE
Introduce "SBOM Type" to the Product essentials 

### ISSUE THREAD

**Issue by DennisClark** [aboutcode-org/dejacode#175]
**Problem**
The SBOM community has identified multiple types of SBOM: 
Software Bill of Materials (SBOM)  -- the default of course
Software-as-a-Service Bill of Materials (SaaSBOM)
Hardware Bill of Materials (HBOM)
Machine Learning Bill of Materials (ML-BOM)
Cryptography Bill of Materials (CBOM)
Manufacturing Bill of Materials (MBOM)
Operations Bill of Materials (OBOM)

This list is ever-expanding; other variants include: 
As-built SBOM
As-deployed SBOM

These various types are implemented variously by different SBOM standards (CycloneDX and SPDX). 

The DejaCode Product definition is quite flexible, and Packages and Components can be defined to identify the kinds of things that exist in each SBOM type, but there is no standard way to identify the SBOM Type associated with a specific DejaCode Product. 

**Benefits**
A new SBOM Type would address that need, and would support additional functionality related to various SBOM Types. Eventually this would support the generation of SBOMs from DejaCode that are more specific than the generic SBOM.

**Comment by mjherzog** [comment]
There is an another useful perspective on SBOM Types at: https://www.linkedin.com/pulse/types-uses-sboms-dirk-riehle-3jcpe/

- SBOMs created from the supplier’s development process
    - Design. A Design SBOM is created from planning documents like prospective product architectures. As a consequence, a Design SBOM may not be an accurate reflection of what will be shipped eventually. It may be helpful to buyers in a supply chain to prepare for what’s to come their way.
    - Source. A Source SBOM provides a static picture of the supplier’s source code and its dependencies, as found in the repositories. It can be helpful to identify vulnerabilities, but does not provide a complete picture as it omits any build or runtime dependencies.
    - Build. A Build SBOM is created from the build process of the supplier as it compiles source code and assembles the final package for delivery to customers. Aimed at operations, it does not include components needed for building and testing. It may still miss dynamic dependencies though.

- SBOMs created by the buyer (or others) through analysis

   -  Analyzed. An Analyzed SBOM is created from software composition analysis of the static delivered software. This is almost always a binary analysis of the artifact. As such, an Analyzed SBOM will miss much, but it may discover components that the suppliers may have overlooked.
    - Deployed. A Deployed SBOM is created from analyzing the deployed software. After deployment, additional components may have been loaded or may have become visible that were not identifiable before. Like Analyzed, Deployed SBOMs complement the supplier’s SBOMs.
   - Runtime. A Runtime SBOM is created from observing the running software (often requiring instrumentation). Of the SBOMs created through analysis, a Runtime SBOM provides the most comprehensive picture, but it will miss components that have not been activated and are not visible yet.

**Comment by DennisClark** [comment]
There's more. SBOMs conforming to SPDX profiles such as:
- Security
- Licensing
- AI
- Data
- Build
- Lite
- Core
- Software

**Comment by DennisClark** [comment]
Here is a link to the CISA definitions for SBOM Types:
https://www.cisa.gov/sites/default/files/2023-04/sbom-types-document-508c.pdf

**Comment by pombredanne** [comment]
I am not sure how an SBOM type would be for a product. A product exists outside of SBOM definitions entirely. Some of its packages and components may have a defined purpose and may be deployed or not. This may be what feeds into these types?

So, if we need to support these SBOM "types", we would need first to define how an inventory can be sliced and categorized.

And before we do this, I would like to see actual real world use case and value from adding this extra layer of complexity, beyond the categorization we already have.

**Comment by mjherzog** [comment]
You are correct that multiple SBOM Types can apply to a Product so we probably need some new data structure that is related to, but separate from Product to handle SBOMs and their Types. See also https://github.com/aboutcode-org/scancode-toolkit/issues/3915.

**Comment by DennisClark** [comment]
Research needed to identify the canonical way to identify an SBOM Type in a generated SBOM for both CycloneDX and SPDX SBOMs.

**Comment by DennisClark** [comment]
Refer to https://docs.google.com/document/d/1_K0qX_IKrfYuezPUp_fVDQIp_tGQ2sEk/edit?usp=sharing&ouid=117241222429542576816&rtpof=true&sd=true    for in-progress design details.

**Comment by nicorikken** [comment]
@DennisClark I get a 401 error on the Google Doc.

I identified the desire to store multiple SBOMs as well: SBOM from version control, SBOM when building code, SBOM when building container, multiple SBOMs from tools that scan the infrastructure (custom software and ready made software).

So not about AI and SaaS, but about different steps in deployment process.

The need stems from a situation of extensive infrastructure with software from different origins. So insights are gathered from different viewpoints, all of which are needed to get a complete insight. It would be different if ALL software would pass through a unified scanning process before deployment.

My ideas about modelling this:

- Stick to common definitions of SBOM type (source, built, ...)
- Use scantool info to make them unique (like 2 different tools scanning container images, so both producing same type of SBOM)
- Have a preferred SBOM type and tool as default view
- Merge component list to a encompassing list of components and set component properties of SBOM tool and SBOM type. These can be used for filtering components.

Alternative is to have duplicate registrations of products: 'mysoftware-source', 'mysoftware-built', 'mysoftware-container', etc.

I'm not a DejaCode user (yet) but this feature is on my want-to-have, which is why I share my perspective.

**Comment by pombredanne** [comment]
@nicorikken the access error is resolved ... try again https://docs.google.com/document/d/1_K0qX_IKrfYuezPUp_fVDQIp_tGQ2sEk/edit
